src/dawn/engine_dawn.cc

```diff
@@ -179,8 +189,253 @@ MapResult MapBuffer(const ::dawn::Device& device, const ::dawn::Buffer& buf) {
   return map_result;
 }
 
+MapResult MapTextureToHostBuffer(const RenderPipelineInfo& render_pipeline,
+                                 const ::dawn::Device& device) {
+  const auto width = render_pipeline.pipeline->GetFramebufferWidth();
+  const auto height = render_pipeline.pipeline->GetFramebufferHeight();
+  const auto pixelSize = render_pipeline.pipeline->GetColorAttachments()[0]
+                             .buffer->GetTexelStride();
+  const auto dawn_row_pitch = Align(width * pixelSize, kMinimumImageRowPitch);
+  const auto fb_buffer_size = dawn_row_pitch * height + width * pixelSize;
+
+  ::dawn::BufferDescriptor descriptor;
+  descriptor.size = fb_buffer_size;
+  descriptor.usage =
+      ::dawn::BufferUsageBit::MapRead | ::dawn::BufferUsageBit::TransferDst;
+  ::dawn::Buffer fb_buffer = device.CreateBuffer(&descriptor);
+
+  ::dawn::TextureCopyView textureCopyView;
+  textureCopyView.texture = render_pipeline.fb_texture;
+  textureCopyView.level = 0;
+  textureCopyView.slice = 0;
+  textureCopyView.origin = {0, 0, 0};
+
+  ::dawn::BufferCopyView bufferCopyView;
+  bufferCopyView.buffer = fb_buffer;
+  bufferCopyView.offset = 0;
+  bufferCopyView.rowPitch = dawn_row_pitch;
+  bufferCopyView.imageHeight = 0;
+
+  ::dawn::Extent3D copySize = {width, height, 1};
+
+  auto encoder = device.CreateCommandEncoder();
+  encoder.CopyTextureToBuffer(&textureCopyView, &bufferCopyView, &copySize);
+
+  auto commands = encoder.Finish();
+  auto queue = device.CreateQueue();
+  queue.Submit(1, &commands);
+
```

Comment:
Maybe wrap lines 201-226 inside a block to make it clear all those things go out of scope at that point?